daml-lf/spec/contract-id.rst

```diff
@@ -0,0 +1,245 @@
+
+This ensures that only the discriminators, not the suffix are needed
+to order the contract IDs of created contract w.r.t. other contract
+IDs. 
+
+Contract ID uniqueness
+----------------------
+
+During interpretation local contract IDs are created without suffix.
+It is the responsibility of the ledger implementation to provide for
+each local contract ID a suffix that guarantee global uniqueness of
+the whole contract ID. No other requirement (except the 95 bytes size
+limit) is assumed for those suffices.
+
+This guard against the submitter choosing low-entropy submission seeds. 
+
+The simplest approach consists to suffix all local contract ID with a
+uniquely global transaction ID. Alternatively central commiter ledger
+can completely avoid suffixing by enforcing that same pair (submission
+seed, submission time) is never used more that once, the discriminator
+allocation scheme ensuring in this case the uniqueness of allocated
+discriminators.
+
+Submission time
+---------------
+
+The submission time is used to derive the transaction seed. In
+practice, it should be closed to the original time at which the
+submission was initiated. No particular requirement is made in this
+specification. However a particular ledger implementation can require
+this time to be within a particular time window around the commit
+time. This allows, for instance, the commiter of a central commiter
+ledger to en enforce global uniqueness of the pair (submission seed,
+submission time).
+
+Allocation scheme for discriminators
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  
+From an initial seed, called in the following *submission seed*, seeds
+for all nodes in the transaction tree are derived as they are
+computed.
+
+We propose the following scheme, inspired by NIST’s HMAC_DRBG
+construction. It assumes a hash-based MAC function HMAC, and defines
+for each big-endian 32 bits integer ``n`` the derived seed::
+
+  deriveSeed(seed, n) := HMAC(seed, n)
+  
+
+Submission seed
+---------------
+
+Any random number with sufficient entropy works as submission seed
+``submission_seed`` (256 bit entropy).
+
+Transaction seed construction
+-----------------------------
+
+From the submission, a so-called *transaction seed* is derived as follows::
+
+  deriveTransactionSeed(submissionSeed, participandId, submissionTime, submitters) :=
+     HMAC(submissionSeed, participantId ∥ submissionTime ∥ nₛ ∥ submitters)
+
+where
+
+* ``submissionSeed`` is the submission seed;
+* ``participandId`` is the participant ID;
+* ``submissionTime`` is the submission time in micro second encoded as
+  a 64 bytes big-endian integer;
+* ``nₛ`` is the number of submitters encoded as a 32 big-endian bits
+  integer;
+* ``submitters`` is the concatenation of the submitter IDs sorted
+  lexicographically.
+
+Strings such as participant IDs or submitter IDs are interpreted as
+their US-ASCII encoding prefixed with there size encoded as a
+32 bits big-endian integer.
```

Comment:
Is the size prefix included when the submitters are sorted lexicographically?

Reply:
Yes. always you get collision for sumittes (`a`, `aaaa`) and (`aa, `aaa`).

Concretely  (`a`, `aaaa`) is encoded as `(2 ∥ 1 ∥ a ∥ 4 ∥ aaaa)` and  (`aa, `aaa`) as  `(2 ∥ 2 ∥ aa ∥ 3 ∥ aaa)`